tests/validate/test_facades.py

```diff
@@ -0,0 +1,51 @@
+# Copyright 2023 Canonical Ltd.
+# Licensed under the Apache V2, see LICENCE file for details.
+
+import importlib
+from collections import defaultdict
+from pathlib import Path
+from typing import Dict, List, TypedDict
+
+import pytest
+
+from juju.client.connection import client_facades, excluded_facades
+
+
+class Versions(TypedDict, total=True):
+    versions: List[int]
+
+
+ClientFacades = Dict[str, Versions]
+
+
+@pytest.fixture
+def project_root(pytestconfig: pytest.Config) -> Path:
+    return pytestconfig.rootpath
+
+
+@pytest.fixture
+def generated_code_facades(project_root: Path) -> ClientFacades:
+    """Return a client_facades dictionary from generated code under project_root.
+    """
+    facades: Dict[str, List[int]] = defaultdict(list)
+    for file in (project_root / 'juju' / 'client').glob('_client[0-9]*.py'):
+        module = importlib.import_module(f'juju.client.{file.stem}')
```

Comment:
```suggestion
    for file in project_root.glob("juju/client/_client*.py"):
        module = importlib.import_module(f'juju.client.{file.stem}')
```

As `.stem` already returns file name without dirs or suffix.

Reply:
Sorry to ask a bunch of questions about this minor suggestion, but I was hoping you could elaborate on a couple of things:
1. Do you prefer `project_root.glob('juju/client/...` over `(project_root / 'juju' / 'client').glob('...` for readability reasons? I prefer the format I used because it seems better to me to do operations explicitly outside of strings where possible, but I'm happy to change this.
2. Do you prefer globbing for `'_client*.py'` rather than `'_client[0-9]*.py'` for terseness, or did you explicitly want to change the glob behaviour? I used `[0-9]` to exclude `_client.py` which doesn't contain any facades. It should still work since we use duck typing to check for facade types. It initially seemed more correct to me to just explicitly check the numbered files, but do you prefer to include `_client.py`?

Reply:
I suppose both, though weakly.

Maybe it's just me, but I tend to think of glob like shell expansion, thus preferring what I see in the shell (and maybe an IDE) over everything-is-an-object.

Same goes for `[0-9]`, like what does this do? how does this cover `_client18`, oh, there's a `*` but not a `+` because it's glob and not regex...

In other words, I think that `project_root.glob("juju/client/_client*.py")` is easier to understand, especially for someone less familiar with the codebase.

Reply:
Makes sense, thanks for explaining. The `[0-9]*` to catch `18` bugged me too, and separately I've been bitten by a glob like that matching too much anyway, so I'll follow your suggestions here.